### PR TITLE
195 Show exif informations

### PR DESCRIPTION
**Pull request by mastermichi**
initial version

### CODE REVIEW

File: res/values-de/strings.xml [hunk 1]
```diff
@@ -78,7 +78,9 @@ Please preserve the xml tags and the utf8 encoding.
     <string name="background_image_duration">Hintergrundbild wechselt alle:</string>
     <string name="seconds_60">60 Sekunden</string>
     <string name="seconds_30">30 Sekunden</string>
-    <string name="seconds_15">15 Sekunden</string>
+    <string name="seconds_15">15 Sekunden</string>#
```
Comment (firebirdberlin) [inline res/values-de/strings.xml:81]:
```suggestion
    <string name="seconds_15">15 Sekunden</string>
```

File: res/values-de/strings.xml [hunk 2]
```diff
@@ -78,7 +78,9 @@ Please preserve the xml tags and the utf8 encoding.
...
+    <string name="slideshow_exif">Zeige EXIF - Informationen</string>
+    <string name="slideshow_exif_summary">Zeigt EXIF - Informationen vom aktuellen Hintergrundbild</string>
```
Comment (firebirdberlin) [inline res/values-de/strings.xml:83]: Please sort by the strings by name witihin the file.

File: res/values/strings.xml [hunk 3]
```diff
@@ -73,6 +73,8 @@ OR
     <string name="seconds_60">60 seconds</string>
     <string name="seconds_30">30 seconds</string>
     <string name="seconds_15">15 seconds</string>
+    <string name="slideshow_exif">Show EXIF - informations</string>
+    <string name="slideshow_exif_summary">Show EXIF - informations from current backgroundimage</string>
```
Comment (firebirdberlin) [inline res/values/strings.xml:77]:
```suggestion
    <string name="slideshow_exif">Show EXIF information</string>
    <string name="slideshow_exif_summary">Show EXIF information for the current background image</string>
```

Please sort by name within the file.

File: src/com/firebirdberlin/nightdream/Settings.java [hunk 4]
```diff
@@ -384,6 +385,7 @@ public void reload() {
         background_zoomin = getBackgroundImageZoomIn();
         background_movein = getBackgroundImageMoveIn();
         background_movein_style = Integer.parseInt(settings.getString("backgroundMovein", "1"));
+        background_exif = settings.getBoolean("slideshowEXIF", false);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/Settings.java:388]:
```suggestion
        background_exif = settings.getBoolean("backgroundEXIF", false);
```

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java [hunk 5]
```diff
@@ -468,6 +477,8 @@ public NightDreamUI(Context context, Window window) {
         background_images[1] = rootView.findViewById(R.id.background_view2);
         background_image_active = 0;
 
+        exif = rootView.findViewById(R.id.exif);
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:480]: exitTextView

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java [hunk 6]
```diff
@@ -718,13 +732,23 @@ private void setupBackgroundImage() {
                         // TODO determine the dominant color in an AsyncTask
                         setDominantColorFromBitmap(preloadBackgroundImage);
                     }
+                    if (settings.background_exif) {
+                        exif.setVisibility(View.VISIBLE);
+                        try {
+                            exif.setText(getExifInformation());
+                        } catch (Exception e) {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:739]: no general exception handlers please

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java [hunk 7]
```diff
@@ -718,13 +732,23 @@ private void setupBackgroundImage() {
...
+                            exif.setText("Error reading exif information");
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:740]: empty on error

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java [hunk 8]
```diff
@@ -744,6 +768,59 @@ private void setupBackgroundImage() {
        }
     }
 
+    private double convertGPS(String[] separated){
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:771]: `convertArcMinToDegrees` ?

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java [hunk 9]
```diff
@@ -744,6 +768,59 @@ private void setupBackgroundImage() {
...
+        String exifCountry = "";
+
+        ExifInterface exif = new ExifInterface(preloadBackgroundImageFile);
+
+        if (exif.getAttribute(ExifInterface.TAG_DATETIME) != null) {
+            String[] exifDateTime = exif.getAttribute(ExifInterface.TAG_DATETIME).split(" ");
+            String[] exifDateSplit = exifDateTime[0].split(":");
+            exifDate = exifDateSplit[2] + "." + exifDateSplit[1] + "." + exifDateSplit[0];
+            exifTime = exifDateTime[1];
+        }
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:799]:
```suggestion
       String datetime = exif.getAttribute(ExifInterface.TAG_DATETIME);
        if (datetime != null) {
            String[] exifDateTime = datetime.split(" ");
            [...] 
        }
```
Ideally settings.dateformat and settings.getFullTimeFormat are used to format date and time.

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java [hunk 10]
```diff
@@ -744,6 +768,59 @@ private void setupBackgroundImage() {
...
+            String[] separatedLat = exif.getAttribute(ExifInterface.TAG_GPS_LATITUDE).split(",");
+            String[] separatedLong = exif.getAttribute(ExifInterface.TAG_GPS_LONGITUDE).split(",");
+
+            Geocoder geocoder = new Geocoder(mContext, Locale.getDefault());
+
+            try {
+                List<Address> addresslist = geocoder.getFromLocation(convertGPS(separatedLat), convertGPS(separatedLong), 1);
+                if (addresslist != null && addresslist.size() > 0) {
+                    exifCity = addresslist.get(0).getLocality();
+                    exifCountry = addresslist.get(0).getCountryName();
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:811]:
```suggestion
                    Address address =  addresslist.get(0);
                    exifCity = address.getLocality();
                    exifCountry = address.getCountryName();
```

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java [hunk 11]
```diff
@@ -744,6 +768,59 @@ private void setupBackgroundImage() {
...
+        ExifInterface exif = new ExifInterface(preloadBackgroundImageFile);
+
+        if (exif.getAttribute(ExifInterface.TAG_DATETIME) != null) {
+            String[] exifDateTime = exif.getAttribute(ExifInterface.TAG_DATETIME).split(" ");
+            String[] exifDateSplit = exifDateTime[0].split(":");
+            exifDate = exifDateSplit[2] + "." + exifDateSplit[1] + "." + exifDateSplit[0];
+            exifTime = exifDateTime[1];
+        }
+
+        if (exif.getAttribute(ExifInterface.TAG_GPS_LATITUDE) != null && exif.getAttribute(ExifInterface.TAG_GPS_LONGITUDE) != null) {
```
Comment (firebirdberlin) [inline src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:801]: extract once and make use of it.

**Review by firebirdberlin — CHANGES_REQUESTED** [review]

File: res/values-de/strings.xml (hunk 1 above)
Comment (mastermichi) [reply res/values-de/strings.xml:81]: done

File: src/com/firebirdberlin/nightdream/Settings.java (hunk 4 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/Settings.java:388]: done

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java (hunk 7 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:740]: done

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java (hunk 10 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:811]: done

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java (hunk 11 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:801]: done

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java (hunk 8 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:771]: ok

File: res/values-de/strings.xml (hunk 2 above)
Comment (mastermichi) [reply res/values-de/strings.xml:83]: done

File: res/values/strings.xml (hunk 3 above)
Comment (mastermichi) [reply res/values/strings.xml:77]: done

File: src/com/firebirdberlin/nightdream/ui/NightDreamUI.java (hunk 5 above)
Comment (mastermichi) [reply src/com/firebirdberlin/nightdream/ui/NightDreamUI.java:480]: done

**Pull request closed without merging**